pkg/kubelet/cm/topologymanager/policy_single_numa_node.go

```diff
@@ -52,8 +54,122 @@ func (p *singleNumaNodePolicy) canAdmitPodResult(hint *TopologyHint) lifecycle.P
 	}
 }
 
+// Merge a TopologyHints permutation to a single hint by performing a bitwise-AND
+// of their affinity masks. At this point, all hints have single numa node affinity
+// and preferred-true.
+func (p *singleNumaNodePolicy) mergePermutation(permutation []TopologyHint) TopologyHint {
+	preferred := true
+	var numaAffinities []bitmask.BitMask
+	for _, hint := range permutation {
+		numaAffinities = append(numaAffinities, hint.NUMANodeAffinity)
+	}
+
+	// Merge the affinities using a bitwise-and operation.
+	mergedAffinity, _ := bitmask.NewBitMask(p.numaNodes...)
+	mergedAffinity.And(numaAffinities...)
+	// Build a mergedHint from the merged affinity mask.
+	return TopologyHint{mergedAffinity, preferred}
+}
+
+// Return hints that have valid bitmasks with exactly one bit set.
+func (p *singleNumaNodePolicy) filterHints(allResourcesHints [][]TopologyHint) [][]TopologyHint {
+	var filteredResourcesHints [][]TopologyHint
+	for _, oneResourceHints := range allResourcesHints {
+		var filtered []TopologyHint
+		if len(oneResourceHints) == 0 {
+			filteredResourcesHints = append(filteredResourcesHints, filtered)
+			continue
+		}
+
+		for _, hint := range oneResourceHints {
+			if hint.NUMANodeAffinity != nil && hint.NUMANodeAffinity.Count() == 1 && hint.Preferred == true {
+				filtered = append(filtered, hint)
+			}
+		}
+
+		filteredResourcesHints = append(filteredResourcesHints, filtered)
+	}
+	return filteredResourcesHints
+}
+
+func (p *singleNumaNodePolicy) mergeProvidersHints(providersHints []map[string][]TopologyHint) TopologyHint {
+	// Loop through all provider hints and save an accumulated list of the
+	// hints returned by each hint provider. If no hints are provided, assume
+	// that provider has no preference for topology-aware allocation.
+	var allResourcesHints [][]TopologyHint
+	for _, hints := range providersHints {
+		if len(hints) == 0 {
+			klog.Infof("[topologymanager] Hint Provider has no preference for NUMA affinity with any resource, " +
+				"skipping.")
+			continue
+		}
+
+		// Otherwise, accumulate the hints for each resource type into allProviderHints.
+		for resource := range hints {
+			if hints[resource] == nil {
+				klog.Infof("[topologymanager] Hint Provider has no preference for NUMA affinity with resource "+
+					"'%s', skipping.", resource)
+				continue
+			}
+
+			if len(hints[resource]) == 0 {
+				klog.Infof("[topologymanager] Hint Provider has no possible NUMA affinities for resource '%s'",
+					resource)
+				// return defaultHint which will fail pod admission
+				return TopologyHint{}
+			}
+			klog.Infof("[topologymanager] TopologyHints for resource '%v': %v", resource, hints[resource])
+			allResourcesHints = append(allResourcesHints, hints[resource])
+		}
+	}
```

Comment:
This all feels very generic (i.e reusable across all policies). Does it make sense to pull this into a top level function as well, something like `gatherAllProviderHints()`?

Reply:
I agree, but I think there are some subtle differences between policies and already a lot of change so I'm inclined to leave it 'as is' for now and maybe revisit later?

Reply:
Fair enough. Sounds good to me.